### PR TITLE
14 create server infos parsing

### LINKED ISSUE #14: Create server infos parsing
Create a parser for informations sent by the server (+ error handling)

For each possible command:
- [ ] Send the command as a string
- [ ] Parse the string response
- [ ] Modify data consequently

### PR DESCRIPTION
**Pull request by Rafouzou**
# Brief

Contains all the methods needed by the client to communicate with the server about the state of the game, both ways (sending and receiving data).

Also contains some new data classes.

Closes #14 .

# Content

1. Commands to send data to the server:

- `Data.updateTeams()` sends the "tna" command to the server (request names of all the teams)
- `Data.updateTime()` sends the "sgt" command to the server (request time unit)
- `Data.editTime()` sends the "sst" command to the server (request time unit modification)
- `Player.updatePosition()` sends the "ppo" command to the server (request player’s position)
- `Player.updateLevel()` sends the "plv" command to the server (request player’s level)
- `Player.updateInventory()` sends the "pin" command to the server (request player’s inventory)
- `Terrain.updateMapSize()` sends the "msz" command to the server (request map size)
- `Terrain.updateTiles()` sends the "mct" command to the server (request content of all the tiles)
- `Tile.updateContent()` sends the "bct" command to the server (request content of one tile)

2. Commands to retrieve data from the server:

- `Data.parseCommand()` parses any command that comes from the server. It is the main entry point.
- `Data.updateTeamsFromString()` parses and manages the "tna" command from the server (name of all the teams)
- `Data.addPlayerFromString()` parses and manages the "pnw" command from the server (connection of a new player)
- `Data.updatePlayerPositionFromString()` parses and manages the "ppo" command from the server (player’s position)
- `Data.updatePlayerLevelFromString()` parses and manages the "plv" command from the server (player’s level)
- `Data.updatePlayerInventoryFromString()` parses and manages the "pin" command from the server (player’s inventory)
- `Data.ejectPlayersFromString()` parses and manages the "pex" command from the server (expulsion)
- `Data.updateBroadcastFromString()` parses and manages the "pbc" command from the server (broadcast)
- `Data.startIncantationFromString()` parses and manages the "pic" command from the server (start of an incantation)
- `Data.endIncantationFromString()` parses and manages the "pie" command from the server (end of an incantation)
- `Data.startEggLayingFromString()` parses and manages the "pfk" command from the server (egg laying by the player)
- `Data.endEggLayingFromString()` parses and manages the "enw" command from the server (resource dropping)
- `Data.dropResourceFromString()` parses and manages the "pdr" command from the server (resource collecting)
- `Data.collectResourceFromString()` parses and manages the "pgt" command from the server (death of a player)
- `Data.killPlayerFromString()` parses and manages the "pdi" command from the server (an egg was laid by a player)
- `Data.connectPlayerToEggFromString()` parses and manages the "ebo" command from the server (player connection for an egg)
- `Data.killEggFromString()` parses and manages the "edi" command from the server (death of an egg)
- `Data.updateTimeFromString()` parses and manages the "sgt" and "sst" commands from the server (time unit update)
- `Data.endGameFromString()` parses and manages the "seg" command from the server (end of game)
- `Data.getServerMessageFromString()` parses and manages the "smg" command from the server (message from the server)
- `Terrain.updateMapSizeFromString()` parses and manages the "" command from the server (map size)
- `Terrain.updateTilesFromString()` parses and manages the "" command from the server (content of all the tiles)

3. New data classes:

- `class Broadcast {
        unsigned int m_player_id;
        std::string m_message;
        unsigned int m_time_left;
}`

The broadcast class contains all the data related to one broadcast

- `class Egg {
        unsigned int m_id;
        unsigned int m_x;
        unsigned int m_y;
}`

The Egg class contains all the data related to one egg

- `class GuiError : public std::exception {
        std::string m_msg;
}`

The GuiError class serves as a special type of exception raised when there is an error with a command that was sent to the gui

- `class ServerMessage {
        std::string m_message;
        unsigned int m_time_left;
}`

The server message class contains all the data related to one server message

### CODE REVIEW

File: gui/src/data.hpp [hunk 1]
```diff
@@ -17,14 +24,63 @@
...
+};
+
+bool is_number(std::string string, unsigned int index);
+std::shared_ptr<Player> find_player(std::vector<std::shared_ptr<Team>> teams, unsigned int id);
+std::shared_ptr<Egg> find_egg(std::vector<std::shared_ptr<Team>> teams, unsigned int id);
+std::shared_ptr<Team> find_team(std::vector<std::shared_ptr<Team>> teams, std::string name);
+std::shared_ptr<Team> find_player_team(std::vector<std::shared_ptr<Team>> teams, unsigned int id);
+std::shared_ptr<Team> find_egg_team(std::vector<std::shared_ptr<Team>> teams, unsigned int id);
+std::shared_ptr<Player> find_player_in_team(std::shared_ptr<Team> team, unsigned int id);
+std::shared_ptr<Egg> find_egg_in_team(std::shared_ptr<Team> team, unsigned int id);
```
Comment (HugoN1) [inline gui/src/data.hpp:86]: Why members and methods outside a class ?

**Comment by Rafouzou** [comment]
Ok actually I changed the 'if' forest to a function pointer map and divided the 'data.cpp' file into 6 different files.

**Review by HugoN1 — CHANGES_REQUESTED** [review]

File: gui/src/data.hpp (hunk 1 above)
Comment (Rafouzou) [reply gui/src/data.hpp:86]: There is no reason for me to put these functions inside the Data class, there is no link between these functions and the Data class.

Comment (AndreasLETANTER) [reply gui/src/data.hpp:86]: Data class is a bad name it can be all or nothing

**Review by HugoN1 — APPROVED** [review]

File: gui/src/data_parsing/data.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+/**
+ * @brief Finds an egg in a list of teams based off its ID
+ * @details Goes through all eggs in all teams and returns the address of the egg we wanted to find, or NULL if it wasn't found
+ * @param Team_address_vector List of teams
+ * @param unsigned_int ID of the egg
+*/
+std::shared_ptr<Egg> find_egg(std::vector<std::shared_ptr<Team>> teams, unsigned int id)
+{
+    for (std::vector<std::shared_ptr<Team>>::iterator team_it = teams.begin(); team_it != teams.end(); ++team_it) {
+        for (std::vector<std::shared_ptr<Egg>>::iterator egg_it = (*team_it)->m_eggs.begin(); egg_it != (*team_it)->m_eggs.end(); ++egg_it) {
```
Comment (lbarreteau) [inline gui/src/data_parsing/data.cpp:130]: auto

File: gui/src/data_parsing/data.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+/**
+ * @brief Finds the team of an egg in a list of teams based off its ID
+ * @details Goes through all eggs in all teams and returns the address of the team of the egg we wanted to find, or NULL if it wasn't found
+ * @param Team_address_vector List of teams
+ * @param unsigned_int ID of the egg
+*/
+std::shared_ptr<Team> find_egg_team(std::vector<std::shared_ptr<Team>> teams, unsigned int id)
+{
+    for (std::vector<std::shared_ptr<Team>>::iterator team_it = teams.begin(); team_it != teams.end(); ++team_it) {
+        for (std::vector<std::shared_ptr<Egg>>::iterator egg_it = (*team_it)->m_eggs.begin(); egg_it != (*team_it)->m_eggs.end(); ++egg_it) {
```
Comment (lbarreteau) [inline gui/src/data_parsing/data.cpp:182]: auto

File: gui/src/data_parsing/data.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * @brief Finds an egg in a team based off its ID
+ * @details Goes through all eggs in the team and returns the address of the egg we wanted to find, or NULL if it wasn't found
+ * @param Team_address Team
+ * @param unsigned_int ID of the egg
+*/
+std::shared_ptr<Egg> find_egg_in_team(std::shared_ptr<Team> team, unsigned int id)
+{
+    for (std::vector<std::shared_ptr<Egg>>::iterator egg_it = team->m_eggs.begin(); egg_it != team->m_eggs.end(); ++egg_it) {
```
Comment (lbarreteau) [inline gui/src/data_parsing/data.cpp:215]: auto

File: gui/src/data_parsing/get_active_player_data.cpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    std::stringstream line_ss(line);
+    line_ss >> command >> x >> y >> result;
+    if (command != "pie") {
+        throw(GuiError("Gui: Data: endIncantationFromString(" + t_str + "): Command '" + command + "' unknown"));
+    } else if (line_ss.fail()) {
+        throw(GuiError("Gui: Data: endIncantationFromString(" + t_str + "): Error parsing command"));
+    }
+    if (result == "ok") {
+        for (std::vector<std::shared_ptr<Team>>::iterator team_it = m_teams.begin(); team_it != m_teams.end(); ++team_it) {
+            for (std::vector<std::shared_ptr<Player>>::iterator player_it = (*team_it)->m_players.begin(); player_it != (*team_it)->m_players.end(); ++player_it) {
```
Comment (lbarreteau) [inline gui/src/data_parsing/get_active_player_data.cpp:81]: auto

File: gui/src/data.hpp (hunk 1 above)
Comment (MATIFIREofficiel) [reply gui/src/data.hpp:86]: Yes Data class is a bad name :/

File: gui/src/data_parsing/get_active_player_data.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                    case S:
+                        if (player_y == m_terrain.getHeight() - 1) {(*player_it)->setY(0); break;}
+                        else {(*player_it)->setY(player_y + 1); break;}
+                    case W:
+                        if (player_x == 0) {(*player_it)->setX(m_terrain.getWidth() - 1); break;}
+                        else {(*player_it)->setX(player_x - 1); break;}
+                }
+            }
+        }
+        for (std::vector<std::shared_ptr<Egg>>::iterator egg_it = (*team_it)->m_eggs.begin(); egg_it != (*team_it)->m_eggs.end(); ++egg_it) {
```
Comment (lbarreteau) [inline gui/src/data_parsing/get_active_player_data.cpp:210]: auto

File: gui/src/data_parsing/get_general_data.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        names_of_new_teams.push_back(team_name);
+    }
+    for (std::vector<std::shared_ptr<Team>>::iterator team_it = m_teams.begin(); team_it != m_teams.end(); ++team_it) {
+        if (std::count(names_of_new_teams.begin(), names_of_new_teams.end(), (*team_it)->getName())) {
+            names_of_kept_teams.push_back((*team_it)->getName());
+        } else {
+            std::vector<std::shared_ptr<Team>>::iterator last_team_it = --m_teams.end();
+            (**team_it) = (**last_team_it);
+            // delete (*last_team_it);
```
Comment (lbarreteau) [inline gui/src/data_parsing/get_general_data.cpp:45]: comment in code

**Review by lbarreteau — CHANGES_REQUESTED** [review]

**Pull request merged**